internal/iam/repository_grant_scope.go

```diff
@@ -0,0 +1,385 @@
+	}
+
+	_, err = r.writer.DoTx(
+		ctx,
+		db.StdRetryCnt,
+		db.ExpBackoff{},
+		func(reader db.Reader, w db.Writer) error {
+			msgs := make([]*oplog.Message, 0, 2)
+			roleTicket, err := w.GetTicket(ctx, &role)
+			if err != nil {
+				return errors.Wrap(ctx, err, op, errors.WithMsg("unable to get ticket"))
+			}
+
+			// We need to update the role version as that's the aggregate
+			updatedRole := allocRole()
+			updatedRole.PublicId = role.GetPublicId()
+			updatedRole.Version = uint32(roleVersion + 1)
+			var roleOplogMsg oplog.Message
+			rowsUpdated, err := w.Update(ctx, &updatedRole, []string{"Version"}, nil, db.NewOplogMsg(&roleOplogMsg), db.WithVersion(&roleVersion))
+			if err != nil {
+				return errors.Wrap(ctx, err, op, errors.WithMsg("unable to update role version"))
+			}
+			if rowsUpdated != 1 {
+				return errors.New(ctx, errors.MultipleRecords, op, fmt.Sprintf("updated role and %d rows updated", rowsUpdated))
+			}
+			msgs = append(msgs, &roleOplogMsg)
+			roleGrantScopesOplogMsgs := make([]*oplog.Message, 0, len(newRoleGrantScopes))
+			if err := w.CreateItems(ctx, newRoleGrantScopes, db.NewOplogMsgs(&roleGrantScopesOplogMsgs)); err != nil {
+				return errors.Wrap(ctx, err, op, errors.WithMsg("unable to add grants"))
+			}
+			msgs = append(msgs, roleGrantScopesOplogMsgs...)
+
+			metadata := oplog.Metadata{
+				"op-type":            []string{oplog.OpType_OP_TYPE_CREATE.String()},
+				"scope-id":           []string{scope.PublicId},
+				"scope-type":         []string{scope.Type},
+				"resource-public-id": []string{role.PublicId},
+			}
+			if err := w.WriteOplogEntryWith(ctx, oplogWrapper, roleTicket, metadata, msgs); err != nil {
+				return errors.Wrap(ctx, err, op, errors.WithMsg("unable to write oplog"))
+			}
+
+			return nil
+		},
+	)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+	roleGrantScopes = make([]*RoleGrantScope, 0, len(newRoleGrantScopes))
+	for _, grantScope := range newRoleGrantScopes {
+		roleGrantScopes = append(roleGrantScopes, grantScope.(*RoleGrantScope))
+	}
+	return roleGrantScopes, nil
+}
+
+// DeleteRoleGrantScopes will delete role grant scopes associated with the role ID in
+// the repository. No options are currently supported. Zero is not a valid value
+// for the WithVersion option and will return an error.
+func (r *Repository) DeleteRoleGrantScopes(ctx context.Context, roleId string, roleVersion uint32, grantScopes []string, _ ...Option) (int, error) {
+	const op = "iam.(Repository).DeleteRoleGrantScopes"
+	if roleId == "" {
+		return db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "missing role id")
+	}
+	if len(grantScopes) == 0 {
+		return db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "missing grant scopes")
+	}
+	if roleVersion == 0 {
+		return db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "missing version")
+	}
+
+	role := allocRole()
+	role.PublicId = roleId
+
+	scope, err := role.GetScope(ctx, r.reader)
```

Comment:
consider moving this read into the Tx just for consistency.  It's not a race issue, because I don't think the role's scope can change.

Reply:
Role's scope can't change; moving this would mean moving getting the oplog wrapper too which I don't think we want to do.